build_runner_core/lib/src/util/build_dirs.dart

```diff
@@ -4,18 +4,27 @@
 
 import 'package:build/build.dart';
 
+import '../generate/options.dart';
 import '../generate/phase.dart';
 
-/// Returns whether or not [id] should be built based upon [buildDirs] and
-/// [phase].
-bool shouldBuildForDirs(AssetId id, List<String> buildDirs, BuildPhase phase) {
+/// Returns whether or not [id] should be built based upon [buildDirs],
+/// [phase], and optional [buildFilters].
+bool shouldBuildForDirs(AssetId id, List<String> buildDirs, BuildPhase phase,
+    {Iterable<BuildFilter> buildFilters}) {
   // Always build non-hidden outputs, since they ship with the package.
   // Otherwise it is too easy to ship not up to date files on accident.
   if (!phase.hideOutput) return true;
 
   // Always build `lib/` dirs, anything in them might be available to any app.
   if (id.path.startsWith('lib/')) return true;
 
+  // Apply the build filters, if present.
+  if (buildFilters?.isNotEmpty == true) {
+    if (!buildFilters.any((f) => f.matches(id))) {
```

Comment:
[nit] Can we use `&&` instead of nested conditions?

Reply:
sure, done